scheduler/core/programprovider/gpp/gppprogramprovider.py

```diff
@@ -0,0 +1,1682 @@
+    def __init__(self,
+                 obs_classes: FrozenSet[ObservationClass],
+                 sources: Sources):
+        super().__init__(obs_classes, sources)
+
+    @staticmethod
+    def parse_notes(notes: Iterable[Tuple[str, str]], search_strings: frozenset) -> bool:
+        """Search note title and content strings
+           Returns a boolean indicating whether the strings were found
+           notes: list of note tuples,  [(title, text), (title, text),...]
+           search_strings: forzenset of strings to search for"""
+
+        # Search for any indications in the note that an observation cannot be split.
+        for note in notes:
+            title, content = note
+            if title is not None:
+                title_lower = title.lower()
+                if any(s in title_lower for s in search_strings):
+                    return True
+            if content is not None:
+                content_lower = content.lower()
+                if any(s in content_lower for s in search_strings):
+                    return True
+        return False
+
+    def parse_magnitude(self, data: dict) -> Magnitude:
+        band = MagnitudeBands[data[GppProgramProvider._MagnitudeKeys.NAME]]
+        value = data[GppProgramProvider._MagnitudeKeys.VALUE]
+        return Magnitude(
+            band=band,
+            value=value,
+            error=None)
+
+    @staticmethod
+    def _get_program_dates(program_type: ProgramTypes,
+                           program_id: ProgramID,
+                           note_titles: List[str]) -> Tuple[datetime, datetime]:
+        """
+        Find the start and end dates of a program.
+        This requires special handling for FT programs, which must contain a note with the information
+        at the program level with key INFO_SCHEDNOTE, INFO_PROGRAMNOTE, or INFO_NOTE.
+        """
+        year_str = program_id.id[3:7]
+        try:
+            year = int(year_str)
+        except ValueError as e:
+            msg = f'Illegal year specified for program {program_id}: {year_str}.'
+            raise ValueError(e, msg)
+        except TypeError as e:
+            msg = f'Illegal type data specified for program {program_id}: {year_str}.'
+            raise TypeError(e, msg)
+        next_year = year + 1
+
+        # Make sure the actual year is in the valid range.
+        if year < 2000 or year > 2100:
+            msg = f'Illegal year specified for program {program_id}: {year_str}.'
+            raise ValueError(msg)
+
+        half_char = program_id.id[7]
+        try:
+            semester = SemesterHalf(half_char)
+        except ValueError as e:
+            msg = f'Illegal semester specified for program {program_id}: {half_char}'
+            raise ValueError(msg, e)
+
+        # Special handling for FT programs.
+        if program_type is ProgramTypes.FT:
+            months_list = [x.lower() for x in calendar.month_name[1:]]
+
+            def is_ft_note(curr_note_title: str) -> bool:
```

Comment:
Same as above this info should rather be in the model

Reply:
We won't be processing GPP notes, all the necessary information has to be in fields.